### DISCUSSION TITLE
Xcode Preview (Canvas tool) Cause Android to Loop restarting the App - Canvas never loads!  Hello Skipper code.

### DISCUSSION THREAD

**Discussion by davidakoontz** [skiptools/skip#274] [Q&A] [answered]
I've attempted to fix the Canvas issue [ Android App continually RESTARTS ] by starting fresh with Hello World (Skipper) codebase.

Before skip init - I upgraded & checked out skip... all Good.
Did the classic > skip init - but with my project names.

Useing the downloaded Skip Hello World code - launch the Canvas - and it hangs... restarting Android in a loop.

Oct 26 2024 - Because of Android restart loop when I launch Xcode Preview (Canvas)… going to start all over.

❯ pwd
/Users/david/Developer/Skip_Projects/circleButtonNav/Android
❯ cd ../..
❯ ls
LetterGame                        circleButtonNav
Wordle-ExploreWordList.playground skip-keychain-main
❯ skip upgrade
[✓] Check Skip Updates: 1.1.15
[✓] Upgrade Skip to 1.1.15 (24.15s)
❯ skip checkup
[✓] Skip version 1.1.15 (= 1.1.15)
[✓] macOS version 15.0.1 (> 13.5.0)
[✓] macOS architecture: ARM
[✓] Swift version 6.0 (> 5.9.0)
[✓] Xcode version 16.0 (> 15.0.0)
[✓] Xcode tools SDKs: 8
[✓] Homebrew version 4.4.2 (> 4.1.0)
[✓] Gradle version 8.10 (> 8.6.0)
[✓] Java version 22.0.2 (> 21.0.0)
[✓] Android Debug Bridge version 1.0.41 (> 1.0.40)
[✓] Android Studio version: 2024.1
[✓] Android tools SDKs: 2
[✓] Resolve dependencies (14.17s)
[✓] Build hello-skip (15.58s)
[✓] Test Swift (9.08s)
[✓] Test Kotlin (66.13s)
[✓] Archive iOS ipa (20.75s)
[✓] Assemble HelloSkip-release.ipa 40 KB
[✓] Verify HelloSkip-release.ipa 40 KB
[✓] Assembling Android apk (150.96s)
[✓] Verify HelloSkip-release.apk 13.9 MB
[✓] Check Swift Package (1.49s)
[✓] Check Skip Updates: 1.1.15
[✓] Skip 1.1.15 checkup succeeded in 280.26s

    ~/Developer/Skip_Projects ···········

Looks Good to Go:  
skip init --appid=com.lifeworksIQ.painless-password ppwd painlesspassword

❯ skip init --appid=com.lifeworksIQ.painless-password ppwd painlesspassword
Initializing Skip application ppwd
[✓] Resolve dependencies (5.57s)
[✓] Build ppwd (14.9s)
[✓] Check Swift Package (0.61s)
[✓] Created module painlesspassword in /Users/david/Developer/Skip_Projects/ppwd
[✓] Skip 1.1.15 init succeeded in 21.12s

https://github.com/user-attachments/assets/accf1196-8624-41b6-9e9e-08671c659c33

**Comment by davidakoontz** [comment]
Should I do something with ENV VAR  ENABLE_PREVIEWS ?

Here is the Run skip gradle script:

```
if [ "${SKIP_ZERO}" != "" ]; then
	echo "note: skipping skip due to SKIP_ZERO"
	exit 0
elif [ "${ENABLE_PREVIEWS}" == "YES" ]; then
	echo "note: skipping skip due to ENABLE_PREVIEWS"
	exit 0
elif [ "${ACTION}" == "install" ]; then
	echo "note: skipping skip due to archive install"
	exit 0
else
	SKIP_ACTION="${SKIP_ACTION:-launch}"
fi
PATH=${BUILD_ROOT}/Debug:${BUILD_ROOT}/../../SourcePackages/artifacts/skip/skip/skip.artifactbundle/macos:${PATH}:${HOMEBREW_PREFIX:-/opt/homebrew}/bin
echo "note: running gradle build with: $(which skip) gradle -p ${PWD}/../Android ${SKIP_ACTION:-launch}${CONFIGURATION:-Debug}"
skip gradle -p ../Android ${SKIP_ACTION:-launch}${CONFIGURATION:-Debug}

```

**Comment by davidakoontz** [comment]
"They" tell me that _HOPE_ is NOT a stock trading plan...  but here I am hoping that an 18.2 upgrade of Xcode will fix my problem.


<img width="1120" alt="Screenshot 2024-10-28 at 1 01 10 PM" src="https://github.com/user-attachments/assets/08fc602f-c9f3-4f99-abc7-48c43e091698">

**Reply by davidakoontz** [reply]
Upgraded to new 18.2 Simulator ...

But did not fix the Skip issue.  Been exploring the issue... here is a WEIRD thing I've found that's unexplained.
The Xcode script to run skip gradle.... 

PATH=${BUILD_ROOT}/Debug:${BUILD_ROOT}/../../SourcePackages/artifacts/skip/skip/skip.artifactbundle/macos:${PATH}:${HOMEBREW_PREFIX:-/opt/homebrew}/bin
**echo "note: running gradle build with**: $(which skip) gradle -p ${PWD}/../Android ${SKIP_ACTION:-launch}${CONFIGURATION:-Debug}"
skip gradle -p ../Android ${SKIP_ACTION:-launch}${CONFIGURATION:-Debug}



In the BUILD LOG:

The LOG does NOT have the proper echo command  **“note: running…”**  the log does NOT include “note:” at beginning!!!

Run custom shell script ‘Run skip gradle’  5.8 seconds

**running gradle build with:**  /Users/david/Library/Developer/Xcode/DerivedData/painlesspassword-bshphqmm….UNIQUE…kedne/Build/Products/../../SourcePackages/artifacts/skip/skip/skip.artifactbundle/macos/skip gradle -p /Users/david/Developer/Skip_Projects/ppwd/Darwin/../Android launchDebug

So it appers the run skip gradle script is stored somewhere else and being used - not the one in the Xcode window!!

This script pulls a skip EXE from DerivedData  then uses a -p parameter with project_dir in Android launchDebug

**Comment by davidakoontz** [comment]
Well ... thinking I was perhaps onto a source of my problem - I stuck a flag "[DAK]" into the run skip gradle script... to see if it was using the one I hoped it was using...  and YES!  Out popped my flag in the build logs...  (no explaination for what happened to the word "note:"...
<img width="1095" alt="Screenshot 2024-10-28 at 2 50 40 PM" src="https://github.com/user-attachments/assets/9a611493-797a-4d4b-a778-f4ace1a3bff6">

**Comment by davidakoontz** [comment]
Is there any chance of help with this issue;  @aabewhite @marcprux  I'd really love to get Previews working again.

**Comment by marcprux** [comment]
Follow-up in https://github.com/orgs/skiptools/discussions/263.

**Comment by davidakoontz** [comment]
Ahh... that helps!  and the workaround for now seems to be:

Xcode > Editor > Canvas (the 2nd item - when a Swift file is active) > **Use legacy preview execution**   <<==  THIS!

**Comment by marcprux** [comment, marked as answer]
Note that the main discussion for this issue can be found at https://github.com/orgs/skiptools/discussions/263
